File: Java/FatorialRecursiva.java
public class FatorialRecursiva {
	
	public static void main(String [] args){
		System.out.println("Fatorial de 0 : " + Fatorial(0) );
		System.out.println("Fatorial de 1 : " + Fatorial(1) );
		System.out.println("Fatorial de 3 : " + Fatorial(3) );
		System.out.println("Fatorial de 7 : " + Fatorial(7) );
	}
	
	public static int Fatorial(int x){
		if(x <= 1)
			return 1;
		else
			return x * Fatorial(x-1);
	}
}

File: src/scala/FatorialRecursiva.scala

def fatorial(n: Long): Long = {
  n match {
    case 0L | 1L => 1L
    case _ => n * fatorial(n - 1L)
  }
}

object Main extends App {
  val data: Map[Long, Long] = Map(
    0L -> 1L,
    1L -> 1L,
    5L -> 120L,
    20L -> 2432902008176640000L
  )
  data.foreach{ (key, value) =>
    val result: Long = fatorial(key)
    assert(result == value)
    println(s"Fatorial($key): $result")
  }
}
